src/DataStructures/Mesh.hpp

```diff
@@ -0,0 +1,149 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines the class template Mesh.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+
+#include "DataStructures/Index.hpp"
+#include "Utilities/Requires.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+namespace Spectral {
+enum class Basis;
+enum class Quadrature;
+}  // namespace Spectral
+/// \endcond
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Holds the number of grid points, basis, and quadrature in each
+ * direction of the computational grid.
+ *
+ * \details A mesh encapsulates all information necessary to construct the
+ * placement of grid points in the computational domain. It does so through a
+ * choice of basis functions, quadrature and number of points \f$N\f$ in each
+ * dimension. The grid points are then identified as the associated collocation
+ * points and can be computed by `Spectral::collocation_points`. This means that
+ * a simulated physical field can be represented by its value on each grid
+ * point and then approximated by a polynomial of degree \f$p=N-1\f$ through a
+ * linear combination of Lagrange polynomials.
+ *
+ * \note A field represented by a `DataVector` has no meaning without an
+ * accompanying `Mesh` to provide context. Only w.r.t a `Mesh` can a
+ * `DataVector` of length `mesh.number_of_grid_points()` be interpreted as the
+ * field values at the collocation points, that are in turn identical to the
+ * nodal expansion coefficients in Lagrange polynomials.
+ *
+ * \tparam Dim the number of dimensions of the computational grid.
+ */
+template <size_t Dim>
+class Mesh : public Index<Dim> {
```

Comment:
after debating Will's question about `number_of_points`, we decided that Mesh should have an `Index<Dim> extents_` instead of deriving from `Index`